tests/zfs-tests/tests/functional/zvol/zvol_misc/zvol_misc_volmode.ksh

```diff
@@ -0,0 +1,287 @@
+		udev_wait
+		[[ -b "$device" ]] && return 0
+	done
+	log_fail "$device does not exist as a block device"
+}
+
+#
+# Verify $device does not exist
+#
+function blockdev_missing # device
+{
+	typeset device="$1"
+
+	# we wait here instead of doing it in a wrapper around 'zfs set snapdev'
+	# because there are other commands (zfs snap, zfs inherit, zfs destroy)
+	# that can affect device nodes
+	for i in {1..3}; do
+		udev_wait
+		[[ ! -e "$device" ]] && return 0
+	done
+	log_fail "$device exists when not expected"
+}
+
+#
+# Verify $property on $dataset is inherited by $parent and is set to $value
+#
+function verify_inherited # property value dataset parent
+{
+	typeset property="$1"
+	typeset value="$2"
+	typeset dataset="$3"
+	typeset parent="$4"
+
+	typeset val=$(get_prop "$property" "$dataset")
+	typeset src=$(get_source "$property" "$dataset")
+	if [[ "$val" != "$value" || "$src" != "inherited from $parent" ]]
+	then
+		log_fail "Dataset $dataset did not inherit $property properly:"\
+		    "expected=$value, value=$val, source=$src."
+	fi
+}
+
+#
+# Create a small partition on $device, then verify if we can access it
+#
+function verify_partition # device
+{
+	typeset device="$1"
+
+	if [[ ! -b "$device" ]]; then
+		log_fail "$device is not a block device"
+	fi
+	# create a small dummy partition: unfortunately we can't use
+	# set_partition() here, becase it internally calls log_must()
+	if is_linux; then
+		parted $device -s -- mklabel gpt || return $?
+		parted $device -s -- mkpart part0 1cyl 1cyl || return $?
+		blockdev --rereadpt $device 2>/dev/null
+	else
+		typeset format_file="$TESTDIR/format_in.$$"
+		echo "partition" >$format_file
+                echo "0" >> $format_file
+                echo "" >> $format_file
+                echo "" >> $format_file
+                echo "1" >> $format_file
+                echo "1m" >> $format_file
+                echo "label" >> $format_file
+                echo "" >> $format_file
+                echo "q" >> $format_file
+                echo "q" >> $format_file
+                format -e -s -d $device -f "$format_file"
+	fi
+	# verify we can access the partition on the device
+	devname="$(readlink -f "$device")"
+	if is_linux; then
+		[[ -b "$devname""p1" ]]
+	else
+		[[ -b "$devname""s0" ]]
+	fi
+	return $?
+}
+
+#
+# Set volmode tunable to $value
+#
+function sysctl_volmode # value
+{
+	typeset value="$1"
+
+	log_note "Setting volmode tunable to $value"
+	if is_linux; then
+		echo "$value" > '/sys/module/zfs/parameters/zvol_volmode'
+	else
+		sysctl 'vfs.zfs.vol.mode' "$value"
+	fi
+	if [[ $? -ne 0 ]]; then
+		log_fail "Unable to set volmode tunable to $value"
+	fi
+}
+
+#
+# Wait for udev to settle, completely.
+# This is quite discomforting, but there's a race condition here
+# (Amazon 2015.09 x86_64 Release (TEST) is good at triggering this)  where the
+# kernel tries to remove zvol device nodes while they're open by [blkid],
+# [zvol_id] or other udev related processes.
```

Comment:
No deadlock, we quite simply can't `zvol_remove_minor_impl()` because some process (`zvol_id` or `blkid`) is keeping it open, even after calling `udevadm settle`. Also because this happens in a callback (`zvol_task_cb` with `op=ZVOL_ASYNC_SET_VOLMODE`) we can't tell userland we failed.

I've added `printk`s to some zvol functions to see this, log here: http://build.zfsonlinux.org/builders/Amazon%202015.09%20x86_64%20Release%20%28TEST%29/builds/2915/steps/shell_8/logs/log

Relevant part:
```
15:04:46.40 [  915.382321] wait_udev_blkid
15:04:46.40 [  915.382405] udevadm trigger
15:04:46.40 [  915.413832] udevadm settle
15:04:46.40 [  915.552605] ZFS: zvol_create_minor_impl(maj=230, min=16)
15:04:46.40 [  915.556116] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=32) [zvol_id] <- testpool/vol
15:04:46.40 [  915.578599] ZFS: <- zvol_create_minor_impl(testpool/volfs/subvol) = 0
15:04:46.40 [  915.578600] ZFS: <- zvol_set_volmode_impl(testpool/volfs/subvol, 1)
15:04:46.40 [  915.578603] ZFS: -> zvol_set_volmode_impl(testpool/volfs, 1)
15:04:46.40 [  915.578603] ZFS: -> zvol_remove_minor_impl(testpool/volfs)
15:04:46.40 [  915.578604] ZFS: <- zvol_remove_minor_impl(testpool/volfs)
15:04:46.40 [  915.578605] ZFS: -> zvol_create_minor_impl(testpool/volfs)
15:04:46.40 [  915.579243] ZFS: <- zvol_create_minor_impl(testpool/volfs) = 22
15:04:46.40 [  915.579244] ZFS: <- zvol_set_volmode_impl(testpool/volfs, 1)
15:04:46.40 [  915.589887] zfs set volmode=none testpool/volfs
15:04:46.40 [  915.598240] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=0) [zvol_id] <- testpool/testvol
15:04:46.40 [  915.598907] ZFS: -> zvol_set_volmode_impl(testpool/volfs/subvol, 3)
15:04:46.40 [  915.598908] ZFS: -> zvol_remove_minor_impl(testpool/volfs/subvol)
15:04:46.40 [  915.598909] ZFS: zvol_remove_minor_impl(testpool/volfs/subvol): zv->zv_open_count = 1
15:04:46.40 [  915.598910] ZFS: zvol_remove_minor_impl() [blkid]
15:04:46.40 [  915.598911] ZFS: <- zvol_remove_minor_impl(testpool/volfs/subvol)
15:04:46.40 [  915.598912] ZFS: <- zvol_set_volmode_impl(testpool/volfs/subvol, 3)
15:04:46.40 [  915.598913] ZFS: -> zvol_set_volmode_impl(testpool/volfs, 3)
15:04:46.40 [  915.598914] ZFS: -> zvol_remove_minor_impl(testpool/volfs)
15:04:46.40 [  915.598915] ZFS: <- zvol_remove_minor_impl(testpool/volfs)
15:04:46.40 [  915.598916] ZFS: <- zvol_set_volmode_impl(testpool/volfs, 3)
15:04:46.40 [  915.611260] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=16) [zvol_id] <- testpool/volfs/subvol
15:04:46.40 [  915.615037] udevadm trigger
15:04:46.40 [  915.678849] udevadm settle
15:04:46.40 [  915.809448] sleep 1
15:04:46.40 [  915.937970] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=32) [zvol_id] <- testpool/vol
15:04:46.40 [  916.064299] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=16) [zvol_id] <- testpool/volfs/subvol
15:04:46.40 [  916.065178] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=0) [zvol_id] <- testpool/testvol
15:04:46.40 [  916.819001] udevadm trigger
15:04:46.40 [  916.848426] udevadm settle
15:04:46.40 [  916.978386] udevadm trigger
15:04:46.40 [  917.081534] udevadm settle
15:04:46.40 [  917.136611] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=32) [zvol_id] <- testpool/vol
15:04:46.40 [  917.205645] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=16) [zvol_id] <- testpool/volfs/subvol
15:04:46.40 [  917.207991] udevadm trigger
15:04:46.40 [  917.230873] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=0) [zvol_id] <- testpool/testvol
15:04:46.40 [  917.301094] udevadm settle
15:04:46.40 [  917.430218] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=32) [zvol_id] <- testpool/vol
15:04:46.40 [  917.454737] sleep 1
15:04:46.40 [  917.492084] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=16) [zvol_id] <- testpool/volfs/subvol
15:04:46.40 [  917.493677] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=32) [zvol_id] <- testpool/vol
15:04:46.40 [  917.501122] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=0) [zvol_id] <- testpool/testvol
15:04:46.40 [  917.517385] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=16) [zvol_id] <- testpool/volfs/subvol
15:04:46.40 [  917.521285] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=0) [zvol_id] <- testpool/testvol
15:04:46.40 [  918.464263] udevadm trigger
15:04:46.40 [  918.508467] udevadm settle
15:04:46.40 [  918.662666] sleep 1
15:04:46.40 [  918.700846] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=0) [zvol_id] <- testpool/testvol
15:04:46.40 [  918.701963] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=32) [zvol_id] <- testpool/vol
15:04:46.40 [  918.714172] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=16) [zvol_id] <- testpool/volfs/subvol
15:04:46.40 [  919.674047] udevadm trigger
15:04:46.40 [  919.708924] udevadm settle
15:04:46.40 [  919.806875] udevadm trigger
15:04:46.40 [  919.855916] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=32) [zvol_id] <- testpool/vol
15:04:46.40 [  919.866986] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=16) [zvol_id] <- testpool/volfs/subvol
15:04:46.40 [  919.897679] udevadm settle
15:04:46.40 [  919.999353] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=0) [zvol_id] <- testpool/testvol
15:04:46.40 [  920.016618] udevadm trigger
15:04:46.40 [  920.097641] udevadm settle
15:04:46.40 [  920.100224] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=16) [zvol_id] <- testpool/volfs/subvol
15:04:46.40 [  920.133078] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=0) [zvol_id] <- testpool/testvol
15:04:46.40 [  920.186866] udevadm trigger
15:04:46.40 [  920.308162] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=32) [zvol_id] <- testpool/vol
15:04:46.40 [  920.341245] udevadm settle
15:04:46.40 [  920.440611] sleep 1
15:04:46.40 [  920.488187] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=16) [zvol_id] <- testpool/volfs/subvol
15:04:46.40 [  920.680094] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=16) [zvol_id] <- testpool/volfs/subvol
15:04:46.40 [  920.688671] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=32) [zvol_id] <- testpool/vol
15:04:46.40 [  920.697264] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=0) [zvol_id] <- testpool/testvol
15:04:46.40 [  920.804245] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=32) [zvol_id] <- testpool/vol
15:04:46.40 [  920.805421] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=0) [zvol_id] <- testpool/testvol
15:04:46.40 [  921.450395] udevadm trigger
15:04:46.40 [  921.476534] udevadm settle
15:04:46.40 [  921.616947] sleep 1
15:04:46.40 [  921.660484] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=0) [zvol_id] <- testpool/testvol
15:04:46.40 [  921.675369] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=32) [zvol_id] <- testpool/vol
15:04:46.40 [  921.678651] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=16) [zvol_id] <- testpool/volfs/subvol
15:04:46.40 [  922.628876] udevadm trigger
15:04:46.40 [  922.676463] udevadm settle
15:04:46.40 [  922.794680] sleep 1
15:04:46.40 [  922.935008] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=16) [zvol_id] <- testpool/volfs/subvol
15:04:46.40 [  922.949494] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=32) [zvol_id] <- testpool/vol
15:04:46.40 [  922.952749] ZFS: zvol_ioctl(BLKZNAME, maj=230, min=0) [zvol_id] <- testpool/testvol
15:04:46.40 =================================================================
15:04:46.40  End of dmesg log
15:04:46.40 =================================================================
15:04:46.40 NOTE: Performing local cleanup via log_onexit (cleanup)
15:04:46.40 lrwxrwxrwx 1 root root  9 Jun 28 15:04 /dev/zvol/testpool/testvol -> ../../zd0
15:04:46.40 lrwxrwxrwx 1 root root 10 Jun 28 15:04 /dev/zvol/testpool/vol -> ../../zd32
15:04:46.40 
15:04:46.40 /dev/zvol/testpool/volfs:
15:04:46.40 total 0
15:04:46.40 lrwxrwxrwx 1 root root 13 Jun 28 15:04 subvol -> ../../../zd16
15:04:46.43 SUCCESS: zfs destroy -r testpool/volfs
15:04:46.44 SUCCESS: zfs destroy -r testpool/vol
15:04:46.46 SUCCESS: zfs inherit volmode testpool
15:04:46.48 SUCCESS: udevadm trigger
15:04:46.57 SUCCESS: udevadm settle
15:04:46.64 /dev/zvol/testpool/volfs/subvol exists when not expected
```

`blkid` is keeping /dev/zd16 (testpool/volfs/subvol) open even after calling `udevadm settle`, so we can't `zvol_remove_minor_impl()` it when we `zfs set volmode=none testpool/volfs`.